sdk/plugin/grpc_backend_server.go

```diff
@@ -14,29 +16,94 @@ import (
 
 var ErrServerInMetadataMode = errors.New("plugin server can not perform action while in metadata mode")
 
+type backendInstance struct {
+	brokeredClient *grpc.ClientConn
+	backend        logical.Backend
+}
+
 type backendGRPCPluginServer struct {
 	pb.UnimplementedBackendServer
 
-	broker  *plugin.GRPCBroker
-	backend logical.Backend
+	broker *plugin.GRPCBroker
 
-	factory logical.Factory
+	instances map[string]backendInstance
+	sync.RWMutex
 
-	brokeredClient *grpc.ClientConn
+	factory logical.Factory
 
 	logger log.Logger
 }
 
+// getBackendInternal returns the backend but does not hold a lock
+func (b *backendGRPCPluginServer) getBackendInternal(ctx context.Context) (logical.Backend, error) {
+	if singleImpl, ok := b.instances["single"]; ok {
+		return singleImpl.backend, nil
+	}
+
+	id, err := pluginutil.GetMultiplexIDFromContext(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	if inst, ok := b.instances[id]; ok {
+		return inst.backend, nil
+	}
+
+	return nil, fmt.Errorf("no backend instance found")
+}
+
+// getBackend holds a read lock and returns the backend
+func (b *backendGRPCPluginServer) getBackend(ctx context.Context) (logical.Backend, error) {
+	b.RLock()
+	defer b.RUnlock()
+	return b.getBackendInternal(ctx)
+}
+
+// getBrokeredClientInternal returns the brokeredClient but does not hold a lock
+func (b *backendGRPCPluginServer) getBrokeredClientInternal(ctx context.Context) (*grpc.ClientConn, error) {
+	if singleImpl, ok := b.instances["single"]; ok {
+		return singleImpl.brokeredClient, nil
+	}
+
+	id, err := pluginutil.GetMultiplexIDFromContext(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	if inst, ok := b.instances[id]; ok {
+		return inst.brokeredClient, nil
+	}
+
+	return nil, fmt.Errorf("no backend instance found")
+}
+
+// getBrokeredClient holds a read lock and returns the backend
+func (b *backendGRPCPluginServer) getBrokeredClient(ctx context.Context) (*grpc.ClientConn, error) {
+	b.RLock()
+	defer b.RUnlock()
+	return b.getBrokeredClientInternal(ctx)
+}
+
 // Setup dials into the plugin's broker to get a shimmed storage, logger, and
 // system view of the backend. This method also instantiates the underlying
 // backend through its factory func for the server side of the plugin.
 func (b *backendGRPCPluginServer) Setup(ctx context.Context, args *pb.SetupArgs) (*pb.SetupReply, error) {
+	var err error
+	id := "single"
```

Comment:
How come we're always fetching the single instance on Setup?

Reply:
Longer answer [here](https://github.com/hashicorp/vault/pull/14946/files#r854490672) but we default to `single` and if it isn't in the map we [call GetMultiplexIDFromContext](https://github.com/hashicorp/vault/pull/14946/files#diff-a5bab53bc70ad6f4380960cdf037117390a05cc089a8d77c21e697eb253c8e2eR95)